File: src/dev/coo2csr.cc
//
// Created by juelin on 12/26/23.
//

#include "coo2csr.h"
#include "../graph/unit_graph.h"
#include <dgl/packed_func_ext.h>
using namespace dgl::runtime;

namespace dgl::dev
{
DGL_REGISTER_GLOBAL("dev._CAPI_COO2CSR")
    .set_body([](DGLArgs args, DGLRetValue *rv) {
      int64_t v_num = args[0];
      NDArray src = args[1];
      NDArray dst = args[2];
      NDArray in_data = args[3];
      bool to_undirected = args[4];
      CHECK_EQ(dst.NumElements(), src.NumElements());
      if (in_data.NumElements() != 0) {
        CHECK_EQ(in_data.NumElements(), src.NumElements());
        const auto [indptr, indices, data] = COO2CSR(v_num, src, dst, in_data, to_undirected);
        *rv=ConvertNDArrayVectorToPackedFunc({indptr, indices, data});
      } else {
        const auto [indptr, indices] = COO2CSR(v_num, src, dst, to_undirected);
        *rv=ConvertNDArrayVectorToPackedFunc({indptr, indices, in_data});
      }
    });
DGL_REGISTER_GLOBAL("dev._CAPI_MakeSym")
    .set_body([](DGLArgs args, DGLRetValue *rv) {
      NDArray in_indptr = args[0];
      NDArray in_indices = args[1];
      NDArray in_data = args[2];
      if (in_data.NumElements() != 0) {
        CHECK_EQ(in_data.NumElements(), in_indices.NumElements());
        const auto [indptr, indices, data] = MakeSym(in_indptr, in_indices, in_data);
        *rv=ConvertNDArrayVectorToPackedFunc({indptr, indices, data});
      } else {
        const auto [indptr, indices] = MakeSym(in_indptr, in_indices);
        *rv=ConvertNDArrayVectorToPackedFunc({indptr, indices, in_data});
      }
    });
    DGL_REGISTER_GLOBAL("dev._CAPI_ReindexCSR")
    .set_body([](DGLArgs args, DGLRetValue *rv) {
      NDArray in_indptr = args[0];
      NDArray in_indices = args[1];
      const auto [out_indptr, out_indices] = ReindexCSR(in_indptr, in_indices);
      *rv=ConvertNDArrayVectorToPackedFunc({out_indptr, out_indices});
    });

    DGL_REGISTER_GLOBAL("dev._CAPI_CompactCSR")
    .set_body([](DGLArgs args, DGLRetValue *rv) {
      NDArray in_indptr = args[0];
      NDArray in_indices_flag = args[1];
      const auto out_indptr = CompactCSR(in_indptr, in_indices_flag);
      *rv=out_indptr;
    });

    DGL_REGISTER_GLOBAL("dev._CAPI_LoadSNAP")
    .set_body([](DGLArgs args, DGLRetValue *rv) {
      std::string in_file = args[0];
      bool to_sym = args[1];
      const auto [out_indptr, out_indices] = LoadSNAP(in_file, to_sym);
      *rv=ConvertNDArrayVectorToPackedFunc({out_indptr, out_indices});
    });
}

File: python/dgl/dev/coo2csr.py
from dgl.backend import to_dgl_nd, from_dgl_nd
from torch import Tensor
from ..heterograph import DGLBlock
from .._ffi.function import _init_api

_init_api("dgl.dev", __name__)

def COO2CSR(v_num: int, src: Tensor, dst: Tensor, data: Tensor=Tensor([]), to_undirected : bool = True):
    src = to_dgl_nd(src)
    dst = to_dgl_nd(dst)
    data = to_dgl_nd(data)
    ret = _CAPI_COO2CSR(v_num, src, dst, data, to_undirected)

    indptr = from_dgl_nd(ret(0))
    indices = from_dgl_nd(ret(1))
    data = from_dgl_nd(ret(2))
    return (indptr, indices, data)

def MakeSym(indptr:Tensor, indices:Tensor, data: Tensor=Tensor([])):
    indptr = to_dgl_nd(indptr)
    indices = to_dgl_nd(indices)
    data = to_dgl_nd(data)
    ret = _CAPI_MakeSym(indptr, indices, data)

    indptr = from_dgl_nd(ret(0))
    indices = from_dgl_nd(ret(1))
    data = from_dgl_nd(ret(2))
    return (indptr, indices, data)

def ReindexCSR(indptr: Tensor, indices: Tensor):
    indptr = to_dgl_nd(indptr)
    indices = to_dgl_nd(indices)
    ret = _CAPI_ReindexCSR(indptr, indices)
    indptr = from_dgl_nd(ret(0))
    indices = from_dgl_nd(ret(1))
    return (indptr, indices)

# compact indptr to remove the edges not in the flag
def CompactCSR(indptr: Tensor, flag: Tensor):
    indptr = to_dgl_nd(indptr)
    flag = to_dgl_nd(flag)
    _indptr = _CAPI_CompactCSR(indptr, flag)
    return from_dgl_nd(_indptr)

def LoadSNAP(in_file: str, to_sym=True):
    ret = _CAPI_LoadSNAP(in_file, to_sym)
    indptr = from_dgl_nd(ret(0))
    indices = from_dgl_nd(ret(1))
    return indptr, indices

def ExpandIndptr(indptr: Tensor):
    src = _CAPI_ExpandIndptr(to_dgl_nd(indptr))
    return from_dgl_nd(src)

def Increment(array: Tensor, row: Tensor):
    return _CAPI_Increment(to_dgl_nd(array), to_dgl_nd(row))